# Введение в глубинное обучение, ФКН ВШЭ

## Домашнее задание 1. Введение в PyTorch. Полносвязные нейронные сети.

### Общая информация

Дата выдачи: 05.10.2021

Мягкий дедлайн: 23:59MSK 19.10.2021

Жесткий дедлайн: 23:59MSK 23.10.2021

Оценка после штрафа после мягкого дедлайна вычисляется по формуле $M_{penalty} = M_{full} \cdot 0.85^{t/1440}$, где $M_{full}$ &mdash; полная оценка за работу без учета штрафа, а $t$ &mdash; время в минутах, прошедшее после мягкого дедлайна (округление до двух цифр после запятой). Таким образом, спустя первые сутки после мягкого дедлайна вы не можете получить оценку выше **8.5**, а если сдать перед самым жестким дедлайном, то ваш максимум &mdash; **5.22** балла.

### Оценивание и штрафы
Максимально допустимая оценка за работу — 10 баллов. Сдавать задание после указанного срока сдачи нельзя.

Задание выполняется самостоятельно. «Похожие» решения считаются плагиатом и все задействованные студенты (в том числе те, у кого списали) не могут получить за него больше 0 баллов. Если вы нашли решение какого-то из заданий (или его часть) в открытом источнике, необходимо указать ссылку на этот источник в отдельном блоке в конце вашей работы (скорее всего вы будете не единственным, кто это нашел, поэтому чтобы исключить подозрение в плагиате, необходима ссылка на источник).

Неэффективная реализация кода может негативно отразиться на оценке.
Также оценка может быть снижена за плохо читаемый код и плохо оформленные графики. Все ответы должны сопровождаться кодом или комментариями о том, как они были получены.

### О задании

В этом задании вам предстоит предсказывать год выпуска песни по некоторым звуковым признакам: [данные](https://archive.ics.uci.edu/ml/datasets/yearpredictionmsd). В ячейках ниже находится код для загрузки данных. Обратите внимание, что обучающая и тестовая выборки располагаются в одном файле, поэтому НЕ меняйте ячейку, в которой производится деление данных.

In [80]:
import torch
from torch import nn
import pandas as pd
import numpy as np
import random
from sklearn.metrics import mean_squared_error
from tqdm.notebook import tqdm
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from torch.nn import functional as F
from torch.utils.data import Dataset, DataLoader
from tqdm import tqdm_notebook
from sklearn.model_selection import train_test_split
from IPython.display import display, clear_output

In [4]:
!wget -O data.txt.zip https://archive.ics.uci.edu/ml/machine-learning-databases/00203/YearPredictionMSD.txt.zip

--2021-10-28 17:25:30--  https://archive.ics.uci.edu/ml/machine-learning-databases/00203/YearPredictionMSD.txt.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 211011981 (201M) [application/x-httpd-php]
Saving to: ‘data.txt.zip’

data.txt.zip        100%[===================>] 201.24M  40.7MB/s    in 5.5s    

2021-10-28 17:25:36 (36.8 MB/s) - ‘data.txt.zip’ saved [211011981/211011981]



In [81]:
df = pd.read_csv('data.txt.zip', header=None)
df.head()

0         1         2         3   ...         87        88         89        90
0  2001  49.94357  21.47114  73.07750  ...   68.40795  -1.82223  -27.46348   2.26327
1  2001  48.73215  18.42930  70.32679  ...   70.49388  12.04941   58.43453  26.92061
2  2001  50.95714  31.85602  55.81851  ... -115.00698  -0.05859   39.67068  -0.66345
3  2001  48.24750  -1.89837  36.29772  ...  -72.08993   9.90558  199.62971  18.85382
4  2001  50.97020  42.20998  67.09964  ...   51.76631   7.88713   55.66926  28.74903

[5 rows x 91 columns]

In [82]:
X = df.iloc[:, 1:].values
y = df.iloc[:, 0].values

train_size = 463715
X_train = X[:train_size, :]
y_train = y[:train_size]
X_test = X[train_size:, :]
y_test = y[train_size:]

## Задание 0. (0 баллов, но при невыполнении максимум за все задание &mdash; 0 баллов)

Мы будем использовать RMSE как метрику качества. Для самого первого бейзлайна обучите `Ridge` регрессию из `sklearn`. Кроме того, посчитайте качество при наилучшем константном прогнозе.

In [83]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

ridge = Ridge()
ridge.fit(X_train, y_train)

y_pred = ridge.predict(X_test)
print('RMSE Ridge =', mean_squared_error(y_pred, y_test, squared=False))

y_pred_best_const = [y_train.mean()]*len(y_test)
print("среднее при best сonst = ",mean_squared_error(y_pred_best_const, y_test, squared=False))


RMSE Ridge = 9.510160711373395
среднее при best сonst =  10.85246390513634


10.931035748764902

## Задание 1. (максимум 10 баллов)

Реализуйте обучение и тестирование нейронной сети для предоставленного вам набора данных. Соотношение между полученным значением метрики на тестовой выборке и баллами за задание следующее:

- $\text{RMSE} \le 9.00 $ &mdash; 4 балла
- $\text{RMSE} \le 8.90 $ &mdash; 6 баллов
- $\text{RMSE} \le 8.80 $ &mdash; 8 баллов
- $\text{RMSE} \le 8.75 $ &mdash; 10 баллов

Есть несколько правил, которых вам нужно придерживаться:

- Весь пайплайн обучения должен быть написан на PyTorch. При этом вы можете пользоваться другими библиотеками (`numpy`, `sklearn` и пр.), но только для обработки данных. То есть как угодно трансформировать данные и считать метрики с помощью этих библиотек можно, а импортировать модели из `sklearn` и выбивать с их помощью требуемое качество &mdash; нельзя. Также нельзя пользоваться библиотеками, для которых сам PyTorch является зависимостью.

- Мы никак не ограничиваем ваш выбор архитектуры модели, но скорее всего вам будет достаточно полносвязной нейронной сети.

- Для обучения запрещается использовать какие-либо иные данные, кроме обучающей выборки.

- Ансамблирование моделей запрещено.

### Полезные советы:

- Очень вряд ли, что у вас с первого раза получится выбить качество на 10 баллов, поэтому пробуйте разные архитектуры, оптимизаторы и значения гиперпараметров. В идеале при запуске каждого нового эксперимента вы должны менять что-то одно, чтобы точно знать, как этот фактор влияет на качество.

- Тот факт, что мы занимаемся глубинным обучением, не означает, что стоит забывать про приемы, использующиеся в классическом машинном обучении. Так что обязательно проводите исследовательский анализ данных, отрисовывайте нужные графики и не забывайте про масштабирование и подбор гиперпараметров.

- Вы наверняка столкнетесь с тем, что ваша нейронная сеть будет сильно переобучаться. Для нейросетей существуют специальные методы регуляризации, например, dropout ([статья](https://jmlr.org/papers/volume15/srivastava14a/srivastava14a.pdf)) и weight decay ([блогпост](https://towardsdatascience.com/weight-decay-l2-regularization-90a9e17713cd)). Они, разумеется, реализованы в PyTorch. Попробуйте поэкспериментировать с ними.

- Если вы чего-то не знайте, не гнушайтесь гуглить. В интернете очень много полезной информации, туториалов и советов по глубинному обучению в целом и по PyTorch в частности. Но не забывайте, что за скатанный код без ссылки на источник придется ответить по всей строгости!

- Если вы сразу реализуете обучение на GPU, то у вас будет больше времени на эксперименты, так как любые вычисления будут работать быстрее. Google Colab предоставляет несколько GPU-часов (обычно около 8-10) в сутки бесплатно.

- Чтобы отладить код, можете обучаться на небольшой части данных или даже на одном батче. Если лосс на обучающей выборке не падает, то что-то точно идет не так!

- Пользуйтесь утилитами, которые вам предоставляет PyTorch (например, Dataset и Dataloader). Их специально разработали для упрощения разработки пайплайна обучения.

- Скорее всего вы захотите отслеживать прогресс обучения. Для создания прогресс-баров есть удобная библиотека `tqdm`.

- Быть может, вы захотите, чтобы графики рисовались прямо во время обучения. Можете воспользоваться функцией [clear_output](http://ipython.org/ipython-doc/dev/api/generated/IPython.display.html#IPython.display.clear_output), чтобы удалять старый график и рисовать новый на его месте.

**ОБЯЗАТЕЛЬНО** рисуйте графики зависимости лосса/метрики на обучающей и тестовой выборках в зависимости от времени обучения. Если обучение занимает относительно небольшое число эпох, то лучше рисовать зависимость от номера шага обучения, если же эпох больше, то рисуйте зависимость по эпохам. Если проверяющий не увидит такого графика для вашей лучшей модели, то он в праве снизить баллы за задание.

**ВАЖНО!** Ваше решение должно быть воспроизводимым. Если это не так, то проверяющий имеет право снизить баллы за задание. Чтобы зафиксировать random seed, воспользуйтесь функцией из ячейки ниже.



In [10]:
def set_random_seed(seed):
    torch.backends.cudnn.deterministic = True
    torch.manual_seed(seed)
    torch.cuda.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)

Вы можете придерживаться любой адекватной струкуры кода, но мы советуем воспользоваться следующими сигнатурами функций. Лучше всего, если вы проверите ваши предсказания ассертом: так вы убережете себя от разных косяков, например, что вектор предсказаний состоит из всего одного числа. В любом случае, внимательно следите за тем, для каких тензоров вы считаете метрику RMSE. При случайном или намеренном введении в заблуждение проверяющие очень сильно разозлятся.

In [84]:
# https://stackoverflow.com/questions/35723472/how-to-use-sklearn-fit-transform-with-pandas-and-return-dataframe-instead-of-num
ss = StandardScaler()
X_train = ss.fit_transform(X_train)
X_test = ss.transform(X_test)

ss2 = StandardScaler()
y_train = ss2.fit_transform(y_train[:, None]).reshape(-1)
y_test = ss2.transform(y_test[:, None]).reshape(-1)

In [ ]:
 МНогую инфу взял с семинара, зачем переделывать если уже хорошо)
https://github.com/Aktsvigun/intro_to_dl_dsba_21_22/blob/b22f9d1402c38e18f33a7ecafb7a9530fc87b998/seminars/seminar_1_solved.ipynb

In [85]:
# Constants
SEED = 42 # random seed for reproducibility
LR = 1e-3 # learning rate, controls the speed of the training
WEIGHT_DECAY = 1e-3 # lambda for L2 reg. ()
NUM_EPOCHS = 15 # num training epochs (how many times each instance will be processed)
GAMMA = 0.9995 # learning rate scheduler parameter
BATCH_SIZE = 32 # training batch size
EVAL_BATCH_SIZE = 300 # evaluation batch size.
DEVICE = 'cpu' #'cuda' # device to make the calculations on

In [86]:
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size = 0.2, random_state = SEED)

In [68]:
class MyDataset(Dataset):
    def __init__(self, X, y):
        super().__init__() # to initialize the parent class
        self.X = X
        self.y = y
        self.len = len(X)

    def __len__(self): # We use __func__ for implementing in-built python functions
        return self.len

    def __getitem__(self, index):
        return self.X[index], self.y[index]

In [70]:
train_set = MyDataset(X_train, y_train)
train_loader = torch.utils.data.DataLoader(
    train_set,
    batch_size = BATCH_SIZE,
    shuffle = True
)

test_set = MyDataset(X_test, y_test)
test_loader = torch.utils.data.DataLoader(
    test_set,
    batch_size = BATCH_SIZE,
    shuffle = True
)

val_set = MyDataset(X_val, y_val)
val_loader = torch.utils.data.DataLoader(
    val_set,
    batch_size = BATCH_SIZE,
    shuffle = True
)

dls = {'train': train_loader, 'val': val_loader, 'test': test_loader}

class Model(nn.Module):
    def __init__(self, in_features = 90, out_features = 1):
        super().__init__()
        self.in_features = in_features
        self.out_features = out_features
        self.hidden_size = 100 # тут заменил, надо поиграть еще int(np.sqrt(in_features + out_features))

        self.sequential = nn.Sequential( # NN architecure, where the modules modify the data sequentially
            nn.Linear(in_features, self.hidden_size), # Linear transformation
            nn.ReLU(), # Activation function 
            nn.Linear(self.hidden_size, self.hidden_size), # Linear transformation
            nn.ReLU(),
            nn.Linear(self.hidden_size, 50), # Linear transformation
            nn.ReLU(),
            nn.Linear(50, self.out_features) # Another Linear transformation

        )

    def forward(self, x): # In the forward function, you define how your model runs, from input to output 
        x = self.sequential(x)
        return x

model = Model()
model.to(DEVICE)

loss_fn = nn.MSELoss() # Loss function, which our model will try to minimize
# Initialize GD method, which will update the weights of the model
optimizer = torch.optim.AdamW(model.parameters(), lr=LR, weight_decay=WEIGHT_DECAY)
# Initialize learning rate scheduler, which will decrease LR according to some rule
criterion = torch.optim.lr_scheduler.ExponentialLR(optimizer, gamma=GAMMA) # тут поменял название, так как в def train
set_random_seed(SEED)


In [71]:
# https://stepik.org/lesson/441757/step/3 так же как и в семе, просто в степике тоже инфа есть для дальнейших функций
def rmse(preds, y_true):
        return ((preds - y_true) ** 2).mean() ** (1/2)

        
def eval_loop(model, X_val, X_test, y_val, y_test, model_is_nn=True, model_name="NN"):
        print(f"Model {model_name}")
        for name, x_eval, y_eval in zip(['val', 'test'], [X_val, X_test], [y_val, y_test]):
            if model_is_nn:
                with torch.no_grad():
                    preds = model(torch.Tensor(x_eval)).view(-1)
            else:
                model.fit(X_train, y_train)
                preds = model.predict(x_eval)
            # Don't forget to rescale the results back to measure loss in original space
            preds_rescheduled = preds * ss2.scale_ + ss2.mean_
            y_eval_rescheduled = y_eval * ss2.scale_ + ss2.mean_
            rmse_val = rmse(preds_rescheduled, y_eval_rescheduled)
            print(f'{name}: RMSE = {rmse_val:.2f}')

**ACHTUNG**
Почему-то некоторые отступы слетают, когда редактируешь ipynb в anaconda, а не в коллабе. Я столько времени убил...

In [89]:
def train(model, optimizer, criterion, train_loader, test_loader):
    '''
    params:
        model - torch.nn.Module to be fitted
        optimizer - model optimizer
        criterion - loss function from torch.nn
        train_loader - torch.utils.data.Dataloader with train set
        test_loader - torch.utils.data.Dataloader with test set
                      (if you wish to validate during training)
    '''
    metrics_dict = {
    "Epoch": [],
    "Train RMSE": [],
    "Val RMSE": [],
  }

    train_losses = []
    train_accuracies = []

  # Train loop
    for epoch in tqdm(range(NUM_EPOCHS)):
      metrics_dict["Epoch"].append(epoch)
      for stage in ['train', 'val']:
          with torch.set_grad_enabled(stage == 'train'): # Whether to start building a graph for a backward pass
              if stage == 'train':
                  model.train()
              else:
                  model.eval()

              loss_at_stage = 0 
              for batch in dls[stage]:
                  x_batch, y_batch = batch
                  x_batch, y_batch = x_batch.to(DEVICE), y_batch.to(DEVICE)

                  y_pred = model(x_batch.float()).view(-1) 
                  loss = loss_fn(y_pred, y_batch.float())
                  if stage == "train":
                      loss.backward() # Calculate the gradients of all the parameters wrt loss
                      optimizer.step() # Update the parameters
                      criterion.step()
                      optimizer.zero_grad() # Zero the saved gradient
                  train_losses += [loss_fn.item()] # for future graphics
                  train_accuracies += [accuracy.item()]
                  with torch.no_grad():
                      loss_at_stage += (torch.square((y_pred - y_batch) * ss2.scale_).sum()).item()
              rmse_at_stage = (loss_at_stage / len(dls[stage].dataset)) ** (1/2)
              metrics_dict[f"{stage.title()} RMSE"].append(rmse_at_stage)
    
      clear_output(wait=True)
      display(pd.DataFrame(metrics_dict))


      #raise NotImplementedError
      

def test(model, criterion, test_loader):
    '''
    params:
        model - torch.nn.Module to be evaluated on test set
        criterion - loss function from torch.nn
        test_loader - torch.utils.data.Dataloader with test set
    ----------
    returns:
        predicts - torch.tensor with shape (len(test_loader.dataset), ),
                   which contains predictions for test objects
    '''
    eval_loop(model, X_val, X_test, y_val, y_test, model_is_nn=True, model_name="NN")
    predicts = torch.ones(len(test_loader.dataset))
    return predicts

In [73]:
train(model, optimizer, criterion, train_loader, test_loader)

Epoch  Train RMSE  Val RMSE
0      0    8.922577  8.825823
1      1    8.689132  8.817060
2      2    8.680944  8.816917

In [74]:
assert test(model, criterion, test_loader).shape[0] == y_test.shape[0]

Model NN
val: RMSE = 8.82
test: RMSE = 8.89


Было
val: RMSE = 8.73
test: RMSE = 8.82 и я не запомнил эту гребанную константу

## Задание 2. (0 баллов, но при невыполнении максимум за все задание &mdash; 0 баллов)

Напишите небольшой отчет о том, как вы добились полученного качества: какие средства использовали и какие эксперименты проводили. Подробно расскажите об архитектурах и значениях гиперпараметров, а также какие метрики на тесте они показывали. Чтобы отчет был зачтен, необходимо привести хотя бы 3 эксперимента.

#1. Изначально модель class Model(nn.Module) по rmse давала около 9.00, а когда я бахнул пару слоев, то стало лучше.
#2. Поигравшись с lR и прочитав во флуд чате стало понятно, что LR около [0.5e-2, 3e-2]. Когда 0.9, то быстро затухает на 8,89 где на 4 эпохе, когда 3e-2, то в целом плохое качество
#3. Так же и со значением weight_decay стало равным 1e-3
#4. не хватает мультипотока, так как обучение происходит не меньше 5 мин, а параметров для ручного изменения много)
#5 в целом self.hidden_size = 100 оптимален, только за счет подкрутки констант, без эвристик можно добится rmse 8,8